File: 2001-2250/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isCritical( int a , int b , int c ){
        if( b>a && b>c ) return true ;
        if( b<a && b<c ) return true ; 
        return false ; 
    }
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        if( head.next == null || head.next.next == null ){
            return new int[]{ -1 , -1 } ; 
        }
        int prev = head.val ; 
        int first_ind = -1 ; 
        int prev_ind = -1 ;
        int d = Integer.MAX_VALUE ; 
        int ind = 1 ; 
        head = head.next ; 
        while( head.next != null ){
            if( isCritical( prev , head.val , head.next.val ) ) {
                if( first_ind == -1 ){
                    first_ind = ind ; 
                }else {
                    d = Math.min( d , ind - prev_ind ) ; 
                }
                prev_ind = ind ; 
            }
            ind++ ; 
            prev = head.val ; 
            head = head.next ; 
        }
        if( d == Integer.MAX_VALUE ){
            return new int[]{ -1 , -1 } ; 
        }
        return new int[]{ d , prev_ind - first_ind } ; 
    }
}

File: 2001-2250/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isCritical( self , a : int , b : int , c : int ) -> bool : 
        if a < b and c < b : 
            return True 
        if a > b and c > b : 
            return True 
        return False 

    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        if not head.next or not head.next.next : 
            return [-1,-1]
        prev = head.val 
        first_ind = -1 
        last_ind = -1 
        head = head.next 
        ind = 1 
        d = pow( 10 , 6 )
        while head.next : 
            if self.isCritical( prev , head.val , head.next.val ) : 
                if first_ind == -1 : 
                    first_ind = ind 
                else :
                    d = min( d , ind - last_ind )
                last_ind = ind 
            ind += 1 
            prev = head.val
            head = head.next 
        
        if last_ind == first_ind : 
            return [-1,-1]
        return [ d , last_ind - first_ind ]
